src/core/net/nat64_translator.cpp

```diff
@@ -133,14 +139,43 @@ Translator::Result Translator::TranslateFromIp6(Message &aMessage)
         ExitNow(res = kNotTranslated);
     }
 
-    mapping = FindOrAllocateMapping(ip6Header.GetSource());
+    switch (ip6Header.GetNextHeader())
+    {
+    case Ip6::kProtoUdp:
+        VerifyOrExit(kErrorNone == aMessage.Read(sizeof(ip6Header), udpHeader),
+                     dropReason = ErrorCounters::kIllegalPacket);
+        srcPortOrId = udpHeader.GetSourcePort();
+        break;
+
+    case Ip6::kProtoTcp:
+        // TODO:
+        break;
+
+    case Ip6::kProtoIcmp6:
+        VerifyOrExit(kErrorNone == aMessage.Read(sizeof(ip6Header), icmp6Header),
+                     dropReason = ErrorCounters::kIllegalPacket);
+        srcPortOrId = icmp6Header.GetId();
+        // TODO: Do not create mappings for ICMP messages that do not need a reply
+        break;
+
+    default:
+        dropReason = ErrorCounters::Reason::kUnsupportedProto;
+        ExitNow(res = kDrop);
+        break;
+    }
+
+    mapping = FindOrAllocateMapping(ip6Header.GetSource(), srcPortOrId, ip6Header.GetNextHeader());
```

Comment:
I am not familiar with the NAT64 details myself, so I may be asking some basic questions:

- Suppose a UDP message passes the translator from a specific `ip6Header.GetSource()` and later a TCP (or ICMPv6 message) comes through from the same device.
- In such a case, the mapping that was added for the initial UDP would be associated with its port number, and on the TCP message, based on the code here, I expect the previous mapping would not be considered a match. Is this the desired behavior?
- I assume we create a new mapping then for each protocol? So the same sender may end up using multiple mappings?

Reply:
Yes, your understanding is correct, we create separate mappings for each different source port/icmp id. We have to do this since we don't have enough IPv4 addresses to create a direct mapping between IPv4 and IPv6. When a reply comes back, we have to have way to distinguish between different sessions. This is basically what any home gateway does to use just a single public IPv4 address. The other difference is that the lifetime of these mappings will be much shorter since it doesn't make sense to keep them for a long time especially for a IMCP Ping. 

Writing this reply I realize that there is also the possibility to have one mapping that contains mapping information for all protocols (udp, tcp, icmp). The problem with this approach is that you could have multiple udp/tcp sessions coming from one devices so you still end up creating multiple mappings per device and they are bigger in size.

Reply:
Yes, I was also thinking of spliting the mapping. Your ideea seems good but it adds some complexity. For now I would leave the current solution and after it is merged I could open a another PR to optimize this part.